### ISSUE TITLE
More specialized desktop and mobile layouts

### ISSUE THREAD

**Issue by bertob** [GeopJr/Tuba#228]
The navigation model is currently not ideal on either desktop or mobile:
- On desktop it's weird to have sidebar items for some things, and view switcher items on the right for others
- On mobile it's weird to have a huge sidebar for what would otherwise be a normal primary menu

Using the new AdwBreakpoints it should be possible to have independent layouts for the two cases, which would help improve the layout for both. Here's a strawman proposal:

- For desktop I'd move everything to the sidebar, so you'd essentially have three new things in there (below Home): Notifications, Conversations, and Search
- For mobile I'd make the sidebar a normal primary menu on the right

**Comment by GeopJr** [comment]
Looks great so far!

Here's a quick attempt at the desktop one (no breakpoints yet) (https://github.com/GeopJr/Tuba/tree/experiment/feat/redesign)

![Screenshot from 2023-05-05 17-02-58](https://user-images.githubusercontent.com/18014039/236480100-225f841d-5ef0-4532-8482-55b5bdeec7c1.png)
![Screenshot from 2023-05-05 17-02-54](https://user-images.githubusercontent.com/18014039/236480108-c011a8d3-0a40-457b-b7b0-ceea31022587.png)

The sidebar needs a complete re-write so I didn't bother for now

(It's probably a bit early to ask but) should the top left sidebar button open the account switcher or the current users' profile and instead put the switcher in the menu next to it?

should the scroll to top button switch to the left side? (it's removed in that branch)
and should the compose / new post button be visible on all views or just home?

**Comment by imhemish** [comment]
Just to say, KDE's Tokodon also somewhat looks like this
![image](https://user-images.githubusercontent.com/74466492/236516559-b849689a-6fc8-42c6-8913-15f9219e8722.png)
Looks awesome!

**Comment by bertob** [comment]
Nice, that's a great start! A few notes:

- The background colors on both panes should be darker, not sure where you'd get the correct values though, cc @exalm 
- The sidebar is too wide, so it looks empty. I'd make it about 220px wide.

I pushed my initial mockup here, including a mobile layout: https://gitlab.gnome.org/Teams/Design/other-app-mockups/-/blob/master/tuba/tuba.png

> should the top left sidebar button open the account switcher or the current users' profile and instead put the switcher in the menu next to it?

I was thinking it'd open the account switcher, but good point about needing to have the link to your own profile somewhere. Maybe it's fine if that's available only from the account switcher?

> should the scroll to top button switch to the left side? (it's removed in that branch)

I was thinking we could try having it be above the new post button.

>  should the compose / new post button be visible on all views or just home?

Just home.

Some open questions, mostly so I don't forget:
- The new account chooser and primary menu
- On mobile, should search be a view in the view switcher or a button in the headerbar?
- Where do the other views go on mobile? I was thinking probably into the primary menu, but could maybe also try a new pattern, e.g. an overview menu at the end of the view switcher, like Elk does:

![image](https://user-images.githubusercontent.com/1908896/236578267-2c95b051-4992-46d3-86ce-842ed1054539.png)

**Comment by Exalm** [comment]
> not sure where you'd get the correct values though,

That's just using new widgets. Really don't do the new styles yet please, it's really hard to get it right without at least toolbar view and you will most likely get it wrong at least somewhere. :) Just keep the raised header, etc and switch over once the new widgets land - libadwaita demo is doing the same thing atm.

**Comment by GeopJr** [comment]
Current state + some more questions:

![image](https://github.com/GeopJr/Tuba/assets/18014039/9189e3f8-e6e9-4f50-b3e6-02f2d41bf5f2)
![image](https://github.com/GeopJr/Tuba/assets/18014039/042e65a9-ea13-446b-8a0d-04fdca1ed3e0)
![image](https://github.com/GeopJr/Tuba/assets/18014039/e2cfbdce-1170-4dee-a84f-4823ec07d99d)
![image](https://github.com/GeopJr/Tuba/assets/18014039/d3cec2a6-b086-4e6b-b209-5476f24980a3)

The compose button will follow the android behavior of 'hide on scroll down' and 'show on scroll up after some distance' (the scroll to top button relocates when the composer button gets revealed or hidden)

In general:

- should the profile handle be ellipsized? the current Tuba release and the mockup do that (angela@gnom...) but after some more thought, I'm not sure if it should. It can be used for impersonation e.g. angela@gnome.org and angela@gnome-fake.example would be shown as angela@gnom... The downside is that it can look a bit weird if both the name and handle are too big 
![image](https://github.com/GeopJr/Tuba/assets/18014039/5b010f9e-a3d7-4c19-85dd-338071fa931a)
- should the home & star icon change to the ones from the mockup?

On desktop:

- should the 'Announcements' and 'Follow Requests' items move to the headerbar menu?
- should the 'Tuba' title be visible on the sidebar?

On mobile:

these are mostly additions to the open questions

- still not entirely sure about the sidebar. If the entries move to the main menu in the headerbar, it's going to be somewhat limiting (no icons or badges, might be overwhelming)
- the bottom bar is an [AdwViewSwitcherBar](https://gnome.pages.gitlab.gnome.org/libadwaita/doc/main/class.ViewSwitcherBar.html), it's not meant to be extended both functionality and style wise. If the search button or a button that opens the sidebar is to be added to the bottom bar, we have to create our own widget instead. Style-wise the closer to the mockup I can do with just CSS is the screenshot below, if we create a custom widget, we will be able to have more control over its layout. The downside is it might not follow future HIG changes automatically or always match AdwViewSwitcherBar (& some other issues).

![image](https://github.com/GeopJr/Tuba/assets/18014039/096f23a4-9d01-45c0-b24e-28374ac1f98c)

Another pattern would be the Google Messages one of top left being the sidebar toggle and top right being the account switcher:
|||
| --- | --- |
| ![Screenshot_20230911-225459_Messages](https://github.com/GeopJr/Tuba/assets/18014039/059c24e7-1c4c-4376-8278-17be07effe72) | ![Screenshot_20230911-225522_Messages](https://github.com/GeopJr/Tuba/assets/18014039/44ac4c25-1676-41cd-b1f9-2f2d0e42f58f) |

**Comment by GeopJr** [comment]
> How about moving them into Notifications? Just like there are Posts, Hashtags, News and For You in Discover or Accounts, Posts and Hashtags in Search.

I have some issues with nesting tabbed views (views with tabs like the main, explore or search one), it will look weird having two switchers for a single view:

(quick edited screenshot mockups, the second switcher will be something else like "All" and "Mentions" for the notifications tab)

![image](https://github.com/GeopJr/Tuba/assets/18014039/6c686062-c4f4-4808-ab88-09da94607664)
![image](https://github.com/GeopJr/Tuba/assets/18014039/8f9d88a2-31ea-46d0-b485-e782441b09ff)

**Comment by LukaszH77** [comment]
OK, but why would you need to have double switchers here? Just show one switcher after clicking Notifications in the Sidebar

![Zrzut ekranu z 2023-09-12 06-12-16](https://github.com/GeopJr/Tuba/assets/121253127/c96fab32-aeb7-4cec-a474-33fd11f7b5c5)

**Comment by bugaevc** [comment]
> should the profile handle be ellipsized? the current Tuba release and the mockup do that (angela@gnom...) but after some more thought, I'm not sure if it should. It can be used for impersonation e.g. angela@gnome.org and angela@gnome-fake.example would be shown as angela@gnom... The downside is that it can look a bit weird if both the name and handle are too big

So I'm using Tuba with two accounts that are both bugaevc@some.domain, and it ellipsizes before the domain name! Given that I have the same avatar on both, I can't distinguish between them at all.

So something should definitely be done about this. That being said letting it break into multiple lines also looks bad.

Could the account switcher perhaps be a popover that's be wider than the sidebar?

**Comment by bugaevc** [comment]
![Screenshot from 2023-09-12 14-40-18](https://github.com/GeopJr/Tuba/assets/10091584/e51fc511-8ca6-4b1e-a9a0-843b0670609f)

This is what it looks like. A few low hanging fruits would be reducing those giant margins and removing the prominent delete button (surfacing the delete functionality somewhere else):
![Screenshot from 2023-09-12 14-39-25](https://github.com/GeopJr/Tuba/assets/10091584/e664ce16-8a99-48d4-a051-6c584deb5102)

And this is what it would look like with the account switcher in a popover:
![Screenshot from 2023-09-12 14-59-36](https://github.com/GeopJr/Tuba/assets/10091584/29cb5512-1d27-4ca8-b310-24d824d1017f)

**Comment by GeopJr** [comment]
> OK, but why would you need to have double switchers here? Just show one switcher after clicking Notifications in the Sidebar

On desktop, yes but on mobile the double switchers cannot be avoided (since on mobile Notifications and Conversations are no longer on the sidebar but instead on a switcher)

> So I'm using Tuba with two accounts that are both bugaevc@some.domain, and it ellipsizes before the domain name! Given that I have the same avatar on both, I can't distinguish between them at all.

My initial question was mostly about the posts on a timeline (where impersonation can happen by ellipsizing the handle). On the account switcher I think the stack looks better than a popover but it would match Fractal. I think ellipsizing is somewhat needed even on the popover for extra long domains.
In the meantime, the rows have a tooltip text with the account handle being hovered:

![image](https://github.com/GeopJr/Tuba/assets/18014039/09cba8d6-dc17-42c9-962e-d5160e8a4909)

**Comment by bertob** [comment]
This is looking great, can't wait for this to land! Is there a branch to try the current state?

> should the home & star icon change to the ones from the mockup?

I think you get the new home icon for free when adwaita icon theme updates

> should the profile handle [in the timeline] be ellipsized?

This seems separate from the navigation changes, so I'd split it out to its own issue

> should the 'Announcements' and 'Follow Requests' items move to the headerbar menu?

Move to the headerbar in which cases? For now I think it's fine to keep them in the sidebar, I think.

> should the 'Tuba' title be visible on the sidebar?

No strong opinion, feel free to keep it for now.

> still not entirely sure about the sidebar. If the entries move to the main menu in the headerbar, it's going to be somewhat limiting (no icons or badges, might be overwhelming)

Hmm? Which entries are moving to the primary menu?

> the bottom bar is an [AdwViewSwitcherBar](https://gnome.pages.gitlab.gnome.org/libadwaita/doc/main/class.ViewSwitcherBar.html), it's not meant to be extended both functionality and style wise. If the search button or a button that opens the sidebar is to be added to the bottom bar, we have to create our own widget instead. Style-wise the closer to the mockup I can do with just CSS is the screenshot below, if we create a custom widget, we will be able to have more control over its layout. The downside is it might not follow future HIG changes automatically or always match AdwViewSwitcherBar (& some other issues).

For now I'd try the following as a next step:
- Keep search in the headerbar on mobile, only move it to the sidebar on desktop
- Keep the view switcher (Home / Notifications / Messages) the same as now on mobile, simply hide it on desktop
- Leave styling changes etc. for later

For the account switcher a popover would be the more standard pattern, and it'd cleanly solve the truncation issue. It'd be good to have some kind of inline check to indicate which account is selected when there's more than one, but that's a separate issue :)

**Comment by LukaszH77** [comment]
> This is looking great, can't wait for this to land! Is there a branch to try the current state?

feat/sidebar/match-moockup-order

**Comment by GeopJr** [comment]
> Move to the headerbar in which cases? For now I think it's fine to keep them in the sidebar, I think.

Sounds good, they weren't in the mockup (not sure if they were in a release then so that checks out) so I assumed they were hidden

> Hmm? Which entries are moving to the primary menu?

All of them (sidebar items) I assume, on mobile, according to the original post "On mobile it's weird to have a huge sidebar for what would otherwise be a normal primary menu"

> For the account switcher a popover would be the more standard pattern, and it'd cleanly solve the truncation issue. It'd be good to have some kind of inline check to indicate which account is selected when there's more than one, but that's a separate issue :)

I'll follow Fractal's design then!

**Comment by GeopJr** [comment]
Is the current progress good enough for the 45 release or should it be postponed until the mobile headerbar & workflow is done / decided?

Summarizing:

| Mockup | Current |
|:---:|:---:|
| ![image](https://github.com/GeopJr/Tuba/assets/18014039/8f2777ad-d95b-41cb-8e9d-9c2020466f89) | ![image](https://github.com/GeopJr/Tuba/assets/18014039/55036abc-e169-4da0-86ca-0dd88a37a86e) |
| Top left open shows the account switcher popover, top right is a menu with all the sidebar items (?) | Top left opens the sidebar |

Also I just noticed two other differences in the mockup:
- Posts no longer have a 3-dot menu button (moved to right click / long touch?)
- Relative time unit is in full (h => hours)

**Comment by bertob** [comment]
Yeah, IMO the current version is more or less ready for a 45 release. I'd still be interested in trying the no-sidebar thing on mobile, but it's a bit experimental to have so much navigation inside the primary menu so it might not work out :)

The 3 dot menu and time unit thing are not essential and can be discussed separately / after the next release.

**Comment by LuNeder** [comment]
Hey! Sorry for reviving this old issue, but may I ask how's the layout chosen? 

On Plasma Mobile, the PC layout seems to be used even with a screen resolution that should be used for mobile. I'm on postmarketOS on qemu, using the flatpak tuba.  
On phosh it shows the mobile layout as expected (tho on phosh I did use the apk version instead, but on plasma the apk version complains about libGLESv2 so I need the flatpak one)

Maybe a setting or command-line flag to force either layout would be nice to add?

**Comment by GeopJr** [comment]
Did this happen with other libadwaita apps (prior to your workaround)?
Unfortunately, I do not have a mobile device to test on but if it happened to all libadwaita (or other GUI toolkits or widget libraries) apps it might be worth it to report it to plasma mobile

FWIW, the layout is based on the window size (for all libadwaita apps that are convergent)

**Comment by LuNeder** [comment]
> Did this happen with other libadwaita apps (prior to your workaround)?

Haven’t tried any other adwaita apps on kde tbh, but I can try and let you know

(also the desktop layout looks really good (on desktop), congrats! (it’s not common to even have a desktop layout on gtk apps these days so tuba really impressed me))

> it might be worth it to report it to plasma mobile

Ye, I’ll also report to plstmarketOS the lack of libGLESv2 as a dependency of tuba in their apk repos

I’ll do some more testing to try to figure out exactly what fixed the layout problem (libGLESv2 or something else that came with phosh) and report back.
